bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -54,7 +53,40 @@ final class BagsManager(executorService: ExecutorService) {
   }
   @volatile private var state: BagsState = BagsState(Set.empty, Set.empty)
 
-  // Setup an LRU cache with max entries BagsManager.CacheSize
+  /** The status of programs and observations being managed by BAGS. */
+  @volatile private var bagsStatuses: Map[SPNodeKey, BagsStatus] = Map.empty
+
+  private def setStatus(key: SPNodeKey, statusOpt: Option[BagsStatus]): Unit = {
+    println(s"+++ setStatus: $key, $statusOpt")
+    bagsStatuses.synchronized {
+        val oldStatusOpt = bagsStatuses.get(key)
+        if (oldStatusOpt != statusOpt) {
+          bagsStatuses = statusOpt match {
+            case Some(status) => bagsStatuses + ((key, status))
+            case None => bagsStatuses - key
+          }
+          notifyBagsStatusListeners(key, oldStatusOpt, statusOpt)
+        }
+      }
+    println(s"+++ setStatus done for $key")
+  }
+
+  private def clearStatusForProgram(prog: ISPProgram): Unit = {
+    bagsStatuses.synchronized {
+        prog.getObservations.asScala.foreach { obs =>
```

Comment:
A small leak will occur here if an observation is deleted before `clearStatusForProgram`, since `prog.getObservations` won't include that observation anymore.  We could get around it in a couple of ways.  A `Map` of `Map`s:  `Map[SPProgramID, Map[SPNodeKey, BagsStatus]]` or maybe just make the key a tuple and filter on keys that contain the program id.  On the other hand, it probably doesn't matter that much.